### PR TITLE
add do...catch...finally

### PR DESCRIPTION
**Pull request by BasedUser**
nuff said

### CODE REVIEW

File: Slang.sldef [hunk 1]
```diff
@@ -115,6 +115,12 @@ forloop: 'for' identifier 'in' expr block (\n* 'else' block)?
 
 whileloop: 'while' expr block (\n* 'else' block)?
 
+docall: 'do' block catchclause* finallyclause?
+
+catchclause: 'catch' type identifier block
+
+finallyclause: 'finally' block
```
Comment (egormanga) [inline Slang.sldef]: These two are non-final.

File: Slang.sldef [hunk 2]
```diff
@@ -115,6 +115,12 @@ forloop: 'for' identifier 'in' expr block (\n* 'else' block)?
 
 whileloop: 'while' expr block (\n* 'else' block)?
 
+docall: 'do' block catchclause* finallyclause?
```
Comment (egormanga) [inline Slang.sldef]: This needs to be added to `statement`s.

File: Slang.sldef [hunk 3]
```diff
@@ -147,15 +153,21 @@ assert: 'assert' expr
 
 super: 'super'
 
+raise: 'raise'
+
+throw: 'throw' expr
+
+resume: 'resume'
```
Comment (egormanga) [inline Slang.sldef]: Please add these to `keywordexpr` as well.

**Review by egormanga — CHANGES_REQUESTED** [review]
LGTM! Thank you.

Still needs to be tested and properly implemented in the AST and at least one of the compilers to be merged.

**Review by egormanga — CHANGES_REQUESTED** [review]
LGTM, waiting for the AST part.

**Comment by BasedUser** [comment]
note: untested as of yet

**Review by egormanga — CHANGES_REQUESTED** [review]
Does it actually parse? Maybe add a test for that.

File: ast.py [hunk 4]
```diff
@@ -380,6 +380,23 @@ class ASTClassAttrgetNode(ASTNode):
...
+	def __str__(self):
+		return f"{self.catch} {self.type_} {self.identifier}{self.block}"
+
+class ASTFinallyClauseNode(ASTNode):
+	finally_: 'finally'
+	block: ASTBlockNode
+	_nl: list['\n'] | None
+
+	def __str__(self):
+		return f"{self.finally_} {self.block}"
```
Comment (egormanga) [inline ast.py]: Actually, no space before the block is required.

File: ast.py [hunk 5]
```diff
@@ -491,6 +513,16 @@ class ASTWhileLoopNode(ASTNode):
...
 
+class ASTDocallNode(ASTNode):
+	do_: 'do'
+	block: ASTBlockNode
+	_nl: list['\n'] | None
+	catch: list[ASTCatchClauseNode] | None
+	finally_: ASTFinallyClauseNode | None
+
+	def __str__(self):
+		return f"""{self.do_} {self.block}{f" {S(' ').join(self.catch)}" if (self.catch) else ''}{f" {self.finally_}" if (self.finally_) else ''}"""
```
Comment (egormanga) [inline ast.py]: ditto

**Review by egormanga — CHANGES_REQUESTED** [review]

**Review by egormanga — APPROVED** [review]

**Pull request merged**